File: csharp/bracket-push/BracketPush.cs
using System;
using System.Collections.Generic;

public static class BracketPush
{
  private static readonly Dictionary<char, char> _brackets = new Dictionary<char, char>
  {
      {'{', '}'},
      {'[', ']'},
      {'(', ')'},
  };

  public static bool IsPaired(string input)
  {
    var stack = new Stack<char>();

    foreach (var ch in input)
    {
      if (_brackets.ContainsKey(ch)) stack.Push(_brackets[ch]);
      else if (_brackets.ContainsValue(ch))
      {
        if (stack.Count == 0) return false;
        if (stack.Pop() != ch) return false;
      }
    }

    return stack.Count == 0;
  }
}

File: fsharp/bracket-push/BracketPush.fs
module BracketPush

let isPaired input =
    let parens = dict [ '(',')'; '[',']'; '{','}' ] 

    let rec check input stack = 
        match input, stack with
        | x::xs, _ when parens.ContainsKey x -> 
            check xs (parens.[x]::stack)
        | x::xs, s::ss when x = s ->
            check xs ss
        | x::_, _ when parens.Values.Contains x ->
            false
        | _::xs, _ -> 
            check xs stack
        | [], [] -> true
        | [], _ -> false

    check (Seq.toList input) []